### DISCUSSION TITLE
What is the difference between oh-my-bash completion and bash-completion?

### DISCUSSION THREAD

**Discussion by madalinignisca** [ohmybash/oh-my-bash#573] [Q&A] [answered]
What is the difference between oh-my-bash completion and bash-completion with each package completion script?

I'm confused if to add or not for example git completion to oh-my-bash, if I already have the git's bash completions enabled in bash-completion?

**Comment by akinomyoga** [comment, marked as answer]
They are just independent implementations and unrelated. I'd say the quality is usually better in `bash-completion`, but some completions in Oh My Bash cover commands not covered by `bash-completion`.

> I'm confused if to add or not for example git completion to oh-my-bash,

There are three different providers here.

- `git`
- `bash-completion`
- `oh-my-bash`

Actually, the `git` completion is provided by `git`. `bash-completion` itself does not provide the `git` completion. `oh-my-bash` doesn't provide the `git` completion either. However, both `bash-completion` and `oh-my-bash` have a mechanism to search and load the `git` completion provided by the `git` installation. Here's the detail:

- The `git` packages of some distributions install a hook into `bash-completion` so that the `git` completion provided by the `git` package is automatically loaded by `bash-completion`. If you observe that the `git` completion is ready just by loading `bash-completion`, it means that the `git` package in your distribution sets it up.
- `oh-my-bash`'s "`git` completion" is actually just a loader of the `git` completion by `git`. It searches the system for any `git` completion provided by the `git` package. When it doesn't find any `git` completion, it falls back to the bundled version of the `git` completion, which is just a copy of the `git` completion in a certain version of the upstream `git` project.

So as far as the `git` completion provided by the `git` package is found in the system, the `git` completion loaded through `bash-completion` and that loaded through `oh-my-bash` should be equivalent. Just the loading locations are different.

> if I already have the git's bash completions enabled in bash-completion?

If the `git` completion is already enabled through the hook in `bash-completion`, actually Oh My Bash's `git` completion becomes no-op because it detects the existing `git` completion and skips the processing.

**Comment by madalinignisca** [comment]
Thank you very much @akinomyoga. That was very detailed.
